### PR TITLE
bpo-34753: Use coroutine object or coroutine function instead of coroutine

### PR DESCRIPTION
**Pull request by Windsooon**
There are quite a lot of them, maybe we can replace it step by step.

https://bugs.python.org/issue34753

### CODE REVIEW

File: Doc/library/asyncio-api-index.rst [hunk 1]
```diff
@@ -19,7 +19,7 @@ await on multiple things with timeouts.
     :class: full-width-table
 
     * - :func:`run`
-      - Create event loop, run a coroutine, close the loop.
+      - Create event loop, run a coroutine function, close the loop.
```
Comment (1st1) [inline Doc/library/asyncio-api-index.rst]: coroutine function -> coroutine object

File: Lib/asyncio/locks.py [hunk 2]
```diff
@@ -57,7 +57,7 @@ def __exit__(self, *args):
 
     @coroutine
     def __iter__(self):
-        # This is not a coroutine.  It is meant to enable the idiom:
+        # This is not a coroutine function.  It is meant to enable the idiom:
```
Comment (1st1) [inline Lib/asyncio/locks.py]: "This is not a coroutine function." -> just remove this sentence, it doesn't make a lot of sense anyways.

File: Lib/asyncio/runners.py [hunk 3]
```diff
@@ -6,9 +6,9 @@
 
 
 def run(main, *, debug=False):
-    """Run a coroutine.
+    """Run a coroutine function.
```
Comment (1st1) [inline Lib/asyncio/runners.py]: coroutine object

File: Lib/asyncio/runners.py [hunk 4]
```diff
@@ -34,7 +34,7 @@ def run(main, *, debug=False):
             "asyncio.run() cannot be called from a running event loop")
 
     if not coroutines.iscoroutine(main):
-        raise ValueError("a coroutine was expected, got {!r}".format(main))
+        raise ValueError("a coroutine object was expected, got {!r}".format(main))
```
Comment (1st1) [inline Lib/asyncio/runners.py:37]: I'd add another check before this one for `coroutines.iscoroutinefunction(main)` with a more specific error message.

**Review by 1st1 — CHANGES_REQUESTED** [review]

File: Doc/library/asyncio-api-index.rst (hunk 1 above)
Comment (Windsooon) [reply Doc/library/asyncio-api-index.rst]: I'm sorry I make a mistake here, I will try to learn more about asyncio before moving forward.

Comment (1st1) [reply Doc/library/asyncio-api-index.rst]: NP, it's OK to make mistakes!

File: Lib/asyncio/runners.py [hunk 5]
```diff
@@ -6,9 +6,9 @@
 
 
 def run(main, *, debug=False):
-    """Run a coroutine.
+    """Run a coroutine object.
 
-    This function runs the passed coroutine, taking care of
+    This function runs the passed coroutine function, taking care of
```
Comment (1st1) [inline Lib/asyncio/runners.py]: coroutine function -> coroutine object

**Comment by csabella** [comment]
@Windsooon, would you be able to resolve the merge conflicts?  It seems there was interest in this PR, so maybe it can be finished up.  Thanks!

**Comment by 1st1** [comment]
Ideally I'd like to not merge this is as this isn't backwards compatible.  We'll likely have TaskGroups in asyncio 3.9 and will simply deprecate `asyncio.gather`.

**Comment by AlexWaygood** [comment]
Closing as OP has been unresponsive for over 3 years, and there is a merge conflict. @Windsooon, if you'd still like to work on this, feel free to open a new PR! Alternatively, ping me, and I'll happily reopen the PR.

It also seems like this one is nearly over the line, so there's a possibility that somebody else could pick it up where @Windsooon left off and get it to a mergeable state (giving @Windsooon credit as a co-author).

**Pull request closed without merging**